File: q0049/q0049.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)
        for s in strs:
            charCount = [0] * 26
            for c in s: charCount[ord(c) - ord('a')] += 1
            res[tuple(charCount)].append(s)
        return res.values()

File: q0049/q0049.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        Map<String, List<String>> groups = new HashMap<>();
        for (String s : strs) {
            // Get the key that is used to group the strings
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);

            // Add the string to the relevant group
            groups.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        res.addAll(groups.values());
        return res;
    }
}